File: backend/app/schemas/url.py
from pydantic import BaseModel, Field, HttpUrl


class ShortenRequest(BaseModel):
    url: HttpUrl = Field(..., max_length=2048)
    custom_alias: str | None = Field(
        default=None,
        min_length=3,
        max_length=40,
        pattern=r"^[a-zA-Z0-9_-]+$",
    )


class ShortenResponse(BaseModel):
    short_url: str
    short_code: str
    original_url: str
    is_custom: bool

File: frontend/src/types/url.ts
export interface ShortenRequest {
    url: string
    custom_alias?: string
}

export interface ShortenResponse {
    short_url: string
    short_code: string
    original_url: string
    is_custom: boolean
}

export interface ValidationError {
    type: string
    loc: string[]
    msg: string
    input: unknown
}

export interface ApiError {
    detail: string | ValidationError[]
}

export type ShortenResult =
    | { success: true; data: ShortenResponse }
    | { success: false; error: string }
